### PR TITLE
CLI `ls` command

### PR DESCRIPTION
**Pull request by hotsphink**
I hadn't noticed the trace_save_dir() weirdness. It came in with 9767272. I fixed it but left it more or less as-is, since I couldn't tell where the right place to expose something like that would be. A static member of StreamWriter, maybe? But I figure there's not much point in me guessing when you can just tell me.

### CODE REVIEW

**Comment by hotsphink** [comment]
Oh, and the code I was fixing was using iostream stuff, so I continued that trend, but perhaps you would prefer sprintf'ing into an on-stack char buffer? Oh well, I'm sure you'll have lots of comments.

**Comment by rocallahan** [comment]
Can you squash these commits and rebase to master?

**Comment by hotsphink** [comment]
I *think* this is what you were requesting. I learned me some git-fu and have much more respect now for its ability to rebase away even merge commits. This one surviving commit contains a mixture of @zmaril's original code and my changes, and is labeled as authored by zmaril (which is fine with me). Hopefully that's what you were looking for.

File: src/LsCommand.cc [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    struct stat st;
+    if (stat(path.c_str(), &st) == -1) {
+      cerr << "stat " << path << " failed\n";
+      return false;
+    }
+
+    bytes += st.st_size;
+  }
+
+  const char suffixes[] = " KMGT";
```
Comment (rocallahan) [inline src/LsCommand.cc]: add `static`

File: src/LsCommand.cc [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+      flags.sort_by_time = true;
+      break;
+    default:
+      assert(0 && "Unknown option");
+  }
+
+  return true;
+}
+
+typedef pair<dirent*, TraceReader*> TraceInfo;
```
Comment (rocallahan) [inline src/LsCommand.cc]: Can we use `unique_ptr` here to express the ownership?

File: src/LsCommand.cc [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    string exe = get_exec_path(t);
+
+    // Record date & runtime estimates
+    struct stat stat_version;
+    struct stat stat_data;
+    string version_file = traces_dir + "/" + t.first->d_name + "/version";
+    string data_file = traces_dir + "/" + t.first->d_name + "/data";
+    stat(version_file.c_str(), &stat_version);
+    stat(data_file.c_str(), &stat_data);
+    long int duration = stat_data.st_mtime - stat_version.st_mtime;
```
Comment (rocallahan) [inline src/LsCommand.cc]: This isn't a most robust way to get the duration. In particular if someone naively copies a trace from one place to another and then examines it with `rr ls`, the duration will be reported as zero, or something very small or even negative. I think that's a problem.

A foolproof way to get the duration would be to read the timestamps from the first and last trace records, but that can be slow for very large traces.

I think for now you should probably just remove the duration display so we can get this landed.

File: src/LsCommand.cc [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+  return true;
+}
+
+static bool is_valid_trace(const string& entry) {
+  if (entry[0] == '.')
+    return false;
+  if (entry[0] == '#')
+    return false;
+  if (entry[entry.length() - 1] == '~')
+    return false;
```
Comment (rocallahan) [inline src/LsCommand.cc]: Put {} around all `if` bodies (or combine these into a single boolean expression

**Comment by rocallahan** [comment]
Thanks!

**Comment by zairil** [comment]
So sweet

**Pull request merged**